pkg/deployment/nova.go

```diff
@@ -73,3 +90,16 @@ func DeployNovaExternalCompute(ctx context.Context, helper *helper.Helper, obj c
 	return ctrl.Result{}, novaExternalCompute, nil
 
 }
+
+// GetNovaTemplate returns the NovaTemplate instance to be used. The NovaTemplate
+// in the OpenStackDataPlaneNode if defined takes precedence over the NovaTemplate
+// in the OpenStackDataPlaneRole.
+func GetNovaTemplate(
+	node *dataplanev1beta1.OpenStackDataPlaneNode,
+	role *dataplanev1beta1.OpenStackDataPlaneRole,
+) *dataplanev1beta1.NovaTemplate {
+	if node.Spec.Node.Nova == nil {
```

Comment:
good

Reply:
Hmm.. I thought we've an agreement to do everything for a role including external compute reconciliation in the dataplanerole[1]. This PR seems to be going in the opposite direction to what I'm doing in https://github.com/openstack-k8s-operators/dataplane-operator/pull/165.  We probably need to discuss and agree on this, else it would be lot of wasted effort. 

[1] https://github.com/openstack-k8s-operators/dataplane-operator/blob/main/pkg/deployment/deployment.go#L348

Reply:
To clarify a bit more, we should decide whether to implement all provisioning/deployment logic in  dataplanerole or dataplanenode reconciliation but not in both.

Reply:
I also missing some context here for sure. :) 

So far I understood that dataplane-operator can deploy individual EDPM nodes via DataplaneNode CR or a set of EDPM nodes via DataplaneRole. Hence my change try to keep both use cases supported with NovaExternalCompute. 

At least the [doc](https://github.com/openstack-k8s-operators/dataplane-operator/blob/main/docs/inheritance.md) still says:
> It's also possible to create a node directly outside of a role CR and define its role. If the following CR is created

Reply:
> E.g. Does it mean that the user will not be allowed to create DataplaneNode CR just Role CR? Or we still allow Node CR to be created but only with a valid Role reference?

I think the current user interface is to create a Dataplane CR (neither role/node directly). But there are some discussions I have seen about not having Dataplane CR at all.  I would probably do away with the DataplaneNode CR if all reconciliation is done with DataPlaneRole using one AnsibleEE job (i.e ansible play) per role.

Reg documented agreements, I have not seen one. My understanding is based on baremetal provisioning meeting discussions where we discussed dataplane-operator CRD design.

Reply:
Even if the end user created Dataplane CR I think it currently still possible to only define node in Dataplane CR and not any role.

https://github.com/openstack-k8s-operators/dataplane-operator/blob/c11eadc0c91a90de6b4e67e0e14098ea7fed3333/api/v1beta1/openstackdataplane_types.go#L32-L35 

https://github.com/openstack-k8s-operators/dataplane-operator/blob/c11eadc0c91a90de6b4e67e0e14098ea7fed3333/api/v1beta1/openstackdataplanenode_types.go#L38-L40

So the user can indirectly create Node CRs that are not referring to any Role CRs. I guess then this will change.

Reply:
Just trying to provide some history on past design plans: 
https://github.com/openstack-k8s-operators/dataplane-operator/blob/main/docs/architecture.md

Reply:
I would use the CRDs as they currently exist as the basis for this patch. We did discuss driving all baremetal provisioning from the role. I don't see how this patch would contradict that? Other that my other inline replies, I'm mostly ok with what is proposed here.

@rabi linked to deployment.go where the common deployment code for NovaExternalComptue now lives, and I think after a rebase, this patch would be more inline with the current deployment patterns. Specifically, this commit:

https://github.com/openstack-k8s-operators/dataplane-operator/commit/e0d8cb1e4aa1b830c26c2236a7568b7fb200c349

The idea is to move any specific deployment logic into deployment.go. The function interfaces there work for a node, a role, or in the special case of NovaExternalCompute either a list of nodes (role deploy) or a list of one node (node deploy).

Once this is rebased, it may alleviate that specific concern.

#### Other comments about the CRD design:

I think what we have now is what we accept as the starting point of the current design. I wouldn't base patches on a different design that is not implemented. If at some point we rework or evolve the design, then of course all impacted code that is already committed would need to be updated.

OpenStackDataPlane allows you to create roles and nodes in a single YAML document, and the created roles/nodes will be owned by the top level DataPlane CR. DataPlane deployment is merely a loop over the roles where we deploy each role. There are no plans to implement a more specialized "DataPlane deployment" (I'm not even sure what that would look like). This is what has been meant when I have said "driving things from the role deployment". It's the role controller that is triggering the role deployment.

A user *can* edit a Role or Node which is created and owned by a DataPlane, however those changes are likely to be overwritten on subsequent reconciliations of DataPlane.

A user *can* create top level Roles and Nodes outside of any DataPlane. This does not necessarily mean they *should* at this point. I think we are still a little ways away from documenting our expected usage paths and best practices.

IMO, and I fully admit I am not k8s API expert, but the patterns in dataplane-operator match other patterns across the k8s API.

E.g., You can create a Deployment, and inline a pod in the YAML document, which creates an owned Pod. You can edit the owned Pod, but your changes would eventually get overwritten. You can also just create top level Pods with their own API and they are not owned by anything. While that's a useful mechanism, it doesn't mean that's how you ought to deploy Pods in production, or that it's a best practice on using k8s.

I tried to model the dataplane-operator CRD design and APIs after these and other patterns I saw in the k8s API (deployStrategy, etc). Of course, that doesn't mean that the k8s API is a perfect API either, and there are a lot of quirks there as well.

Reply:
> I don't see how this patch would contradict that?

It would, as we should either do all reconciliation (provisioning/deployment) of nodes in role or the node controller. As per what you're saying provisioning would be role controller managed but deployment should be managed by both does not sound like a good design to me.

> @rabi linked to deployment.go where the common deployment code for NovaExternalComptue now lives, and I think after a rebase, this patch would be more inline with the current deployment patterns.

I think I'm still confused with the current design. I'm not really concerned about where the code lives, but the actual design around object ownership, reconciliation etc. Let me try and explain a few issues I see.

-  I see deployment.Deploy() is called both from the node and role controller. That means both role/node would try to own[1] (and one would fail) the same NovaExternalCompute.

-  Both role and node controller would try to reconcile NovaExternalCompute at the same time for any changes in it[2].

If we do all baremetal provisioning in role controller, with node CR reconciliation we assume the node has already been provisioned/pre-provisioned?  What's the significance of managed:true for node CRD then? Does that mean node controller has to check if a provisioned BMH exists and/or ssh  to the node before starting reconciliation of a node? 

 

[1] https://github.com/openstack-k8s-operators/dataplane-operator/blob/main/pkg/deployment/nova.go#L60

[2] https://github.com/openstack-k8s-operators/dataplane-operator/blob/main/controllers/openstackdataplanenode_controller.go#L267
https://github.com/openstack-k8s-operators/dataplane-operator/blob/main/controllers/openstackdataplanerole_controller.go#L338

Reply:
there is __NOT__ general agreement on doing all the deployment from a role and I would not want to block this from progressing on that.

i don't belive that driving the deployment of all service at the role level is correct 

i would personally prefer if all reconsiltaiton was driven by the node controller or if we split between the two depending on the where the reosuce is defined.

i.e. where the nova template is defined on the role then it can be created and manage by the role controller and where its defined on the node it should be driven by the node controller.

in the current dataplane design roles are entirely optional.

that is why we have examples of using nodes standalone 
https://github.com/openstack-k8s-operators/dataplane-operator/blob/main/config/samples/dataplane_v1beta1_openstackdataplanenode_edpm_compute_0_full.yaml

roles are just a grouping mechanism for managing stes of datapalneNodes and propagating common settings.

they in turn allow you to execuat e ansible across multiple host via generating a merged inventory but that could be removed entirely and all execution could be drive at the node level if desired.

i don't think we want to exclusively drive things at the node level but we certenly never agree to only drive things at the role level.

to your questions regarding baremental provisioning.

if managed=false i.e. pre-provisoned hosts (which is expected to be our default for the dev preview) the AEE created by the node reconsiler will simply fail if the node is not provisioned. this would be reflected in the ready status and the admin would have to retriger the reconsile by modifying a filed on the CR or similar. i.e. update the hash in the status to for ce the reconsile to happen again.

we currently don't have any kind fo healtcheck of the node to know if its provisioned but
the expectation is that when using manged=false the OpenstackDataplaneNode CR is not created until the requirements for ansible to be able to connect have been fulfilled.

That is also a prerequisite for managed=true.
in that case the OpenstackDataplenNode CRs should not be created by the role controller until the baremetal operator has finsied provisioning the OS on the physical hosts.

so the precondition for creating an OpenStack dataplane Node is ensuring its provsioned with the required ssh public key added to the authorised keys file on ideally a non root use with passwordless sudo.

Reply:
@slagle im not convience we shoudl have OpenStackDataPlane CRD at all and i think we should only be creating the OpenstackDAtaplaneRoles and OpenstackDAtaplaneNode CRs

we add the top level OpenStackDataPlane CRD for not very good reason just to have a top level entity to model the over all datapalne but i don't actually think it add any value and in fact i think its very problematic for larger deployments with 1000 of compute nodes.


The design that gibi has implemented in this PR was the design we agreed to originally in the design doc we created for the interaction between service operator and the dataplane operator. its also what we discussed in brno. i.e. OpenStack services that are deployed on dataplane nodes are modelled by strongly typed service Temaplate the existing of which consoles the deployment of the service. 

This is partly why the AnibleVArs files was intentionally not part of fo the dataplne crd design and why i was so shocked by the fact you added it and called it out in the Q1 demo.
form my perspective groping everything at the role level and exposing the use of ansible for generic service config in the API provide dby the CRD were never agreed to as design goal before they were implemented in the dataplen operator.

i think embdeing 1000s of compute nodes in a single top-level CRD is an anti-pattern that we should not be relying on and i would be very concerned about the debuggability, scalability and user experience to taking that approach.


taking your deployment analagoy the nova template is basically a like a pod spec 
so we are proposing inline a spec at the node/role level to allow the create of the NovaExternalComptue CR. as with k8s pod they can be created in one of several ways

you can create free standing pods, or you can create pods that are part of a a higher level crd i.e. replicaset, deployment, statefulset, deamonset, job
each higher level construct adds additional semantic over the pod wich is the standard and minimum unit of scheduling in k8s.

you can create NovaExternalComptue CR as free-standing resources. for example if you were trying to deploy without the dataplene operator for some reason you just need to provide a valid inventory. i don't really expect that to be used out side of our development and testing but its something we support in the nova operator intentionally to make sure we have no depency on the datapalne operator outside fo the configmap that holds the inventory and the secret that provides the ssh key both of which you can create by hand.

you can also create it as managed by a higher-level resource.
in the current design we to entities that can own it it can either be owned by the OpenStackDataplaneNode (in my original proposal it was always going to be owned by this)
or the OpenstackDataplaneRole.

when we originally discussed the interaction between dataplane role and node i was expecting the role reconsiler to mealy mange creating nodes and the node reconsile would do 99% of the work.

that design evolved to have effectively the same functionality both at the role and node level however not without downsides.
the ability to report errors on a per host per task basis for the openstack dataplen role is much less debugable then when the same ansibel roels are executed purely at the openstackdataplennode level.

in the initial conversation we had on the design i was promoting reporting the status fo the deployment on any given node viat the status fo the openstackDataplaneNode status field.
while the conditions partly do that we don't actually have granular reporting fo the AEE progress there.

we also don't have that for the NovaExternalComptue to be fair but that's partly because the cannot get that info easily formteh AEE CR and i belive its wrong to just use embed the output of the underly in k8s job created by the AEE operator from the nova operator because we should not know about the internal fo how that operator works.
we should only use the public filed on the AEE CRD.

so I think the design that gibi has proposed is consistent with k8s API design in its current form and its also the design pattern we had previously agreed to follow for this integration.

Reply:
> roles are just a grouping mechanism for managing stes of datapalneNodes and propagating common settings.

Are we saying role labeled nodes would be reconciled in role controller and the ones not labeled with role would be reconciled in node controller?  NovaExternalCompute is either owned by Role or Node? It can't be both right? What happens when a role is deleted? What happens to the labelled nodes? We would unprovision the nodes as the BaremetalSets are owned by the role. What would happen to the status of dataplane node CRDs that represent those nodes?

> so the precondition for creating an OpenStack dataplane Node is ensuring its provsioned with the required ssh public key added to the authorised keys file on ideally a non root use with passwordless sudo.

IMO, either we reconcile the node completely from nothing to provisioned/deployed node in node controller based on node spec, it can use shared config from the role (config) it's labelled with or we do everything in the role, probably not in both.

Reply:
if we are unprovisioning the physical servers because the baremental set was release when the role was delete i would also expect the OpenStackDataplaneNode CR for that physical server to be deleted.

assuming we want to allow OpenstackDataplaneRole CRs to be deleted if there exist an OpenstackDataplaneNode that refences them.

i kind fo think we should be using a finalise to prevent that.

i.e. the node controller should create a finalise for each node on the roles if a OpenstackDataplaneNode refences a roles.

i think we want to be more intentional especially in the managed=True case to make sure one typo does not nuke your dataplane hosts.

the workflow i think would make more sense would be to scalne the BaremetalSets to 0 before allowing the delete or as i said enforcing no OpenstackDataplaneNodes reference it.
both would have the same effect at the end fo the day.

but yes right now the  NovaExternalCompute should be owned either by the role or node but not both. if it was own by only one i think the  NovaExternalCompute is best aligned with the OpenstackDataplaneNode but we support both right now.

Reply:
> > I don't see how this patch would contradict that?
> 
> It would, as we should either do all reconciliation (provisioning/deployment) of nodes in role or the node controller. As per what you're saying provisioning would be role controller managed but deployment should be managed by both does not sound like a good design to me (with whatever little I know about k8s).

I don't agree it has to be all one way or another. There exist several patterns throughout the API of this exact thing, but I'll go back to Deployment/Pod. You can create a Pod, or you can create a Deployment which creates a Pod (or set of pods). In one case the Pods are owned by the Deployment, in the other they are not. There are best practices and recommendations for using either approach.

> 
> > @rabi linked to deployment.go where the common deployment code for NovaExternalComptue now lives, and I think after a rebase, this patch would be more inline with the current deployment patterns.
> 
> I think I'm still confused with the current design. I'm not really concerned about where the code lives, but the actual design around object ownership, reconciliation etc. Let me try and explain a few issues I see.
> 
>     * I see deployment.Deploy() is called both from the node and role controller. That means both role/node would try to own[1] (and one would fail) the same NovaExternalCompute.

I disagree that's possible with the code we have. Have you seen this happen? I'd consider that a bug if so. When you deploy a role, that's not's going to trigger the reconciliation loop for a node directly. While a node could be reconciled by some other factor, unless deploy=true, it won't call that function and create another NovaExternalCompute.

Of course it is possible create a series of malformed API requests that get things into a messed up state. That exists everywhere throughout the k8s API though. We should add checks to protect where we can (don't deploy a node if the role deploy is in progress, etc).

> 
>     * Both role and node controller would try to reconcile NovaExternalCompute at the same time for any changes in it[2].

No. That's not what that code means in [2]. The controller reconcile would only be triggered for the owner (either node or role). Again, someone could trigger both at the same time unintentionally, and we should protect against that.

> If we do all baremetal provisioning in role controller, with node CR reconciliation we assume the node has already been provisioned/pre-provisioned? What's the significance of managed:true for node CRD then? Does that mean node controller has to check if a provisioned BMH exists and/or ssh to the node before starting reconciliation of a node?

The managed field was added before baremetal support. As you're working on the baremetal support, I'd propose to remove it if you don't think it's needed. I'd agree it's not based on what you've proposed for baremetal support.

Reply:
>  You can create a Pod, or you can create a Deployment which creates a Pod (or set of pods)

Are we sure it's an apple-to-apple comparison? I'm not aware of a pod controller in k8s.  There are ReplicaSet and Deployment (an higher level abstraction) controllers.  What I can see is that deployment reconcile calls replicaset update/reconcile[1] and does not do things that replicaset has to do. There are also lots of checks/balances around ownership, claiming objects etc. Not sure if we would be able to go that level with our implementation here.

[1] https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/deployment/sync.go#L426

>   Have you seen this happen? I'd consider that a bug if so. 

Yes, I've seen that happen when deployStrategy/deploy is set true for node and role at the same time. Though we can have checks to ensure that they're not set to true at the same time.  I would guess those validations would be complex   when you have number of roles and nodes and node roles are toggled/removed. 

> The controller reconcile would only be triggered for the owner (either node or role).

That's right, they would only watch for events for the objects they own. What do we expect to happen when a role label (from spec) is removed from a node? With the current design role reconcile would tear down everything (including unprovisioning the node) as role owns all objects like NovaExternalCompute. Is that what we expect because  there is still a node object with (deployStrategy/deploy: false)?

> Right, I'd also agree there is not complete agreement. It was somewhat of a compromise, but relatively easy to implement, so I think it's a fair compromise to make.

I had the impression (after the baremetal provisioning calls) that we agreed to do everything with role reconciliation, but it seems I misunderstood. Also, I'm not sure we should do compromises because things are easy to implement. I've concerns around the current design and how we want to add baremetal provisioning to role reconcile, but it seems not everyone is bothered the same way. I rest my case.

Reply:
> > You can create a Pod, or you can create a Deployment which creates a Pod (or set of pods)
> 
> Are we sure it's an apple-to-apple comparison? I'm not aware of a pod controller in k8s. There are ReplicaSet and Deployment (an higher level abstraction) controllers. What I can see is that deployment reconcile calls replicaset update/reconcile[1] and does not do things that replicaset has to do. There are also lots of checks/balances around ownership, claiming objects etc. Not sure if we would be able to go that level with our implementation here.

Yes, you absolutely can create a Pod on its own. https://kubernetes.io/docs/concepts/workloads/pods/

It's not apples to apples. It's a comparison of an API pattern.

> 
> [1] https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/deployment/sync.go#L426
> 
> > Have you seen this happen? I'd consider that a bug if so.
> 
> Yes, I've seen that happen when deployStrategy/deploy is set true for node and role at the same time. Though we can have checks to ensure that they're not set to true at the same time. I would guess those validations would be complex when you have number of roles and nodes and node roles are toggled/removed.

I'll add a check for that.

> 
> > The controller reconcile would only be triggered for the owner (either node or role).
> 
> That's right, they would only watch for events for the objects they own. What do we expect to happen when a role label (from spec) is removed from a node? With the current design role reconcile would tear down everything (including unprovisioning the node) as role owns all objects like NovaExternalCompute. Is that what we expect because there is still a node object with (deployStrategy/deploy: false)?

We don't remove role labels from nodes. If you delete the role, it deletes everything it owns, including the node.

> > Right, I'd also agree there is not complete agreement. It was somewhat of a compromise, but relatively easy to implement, so I think it's a fair compromise to make.
> 
> I had the impression (after the baremetal provisioning calls) that we agreed to do everything with role reconciliation, but it seems I misunderstood. Also, I'm not sure we should do compromises because things are easy to implement. I've concerns around the current design and how we want to add baremetal provisioning to role reconcile, but it seems not everyone is bothered the same way. I rest my case.

All of the deployment is driven from the role reconciliation. It's the role controller that (eventually) creates the OpenStackAnsibleEE CR's. Even for NovaExternalCompute, we deploy the roles, which deploys NovaExternalCompute for the nodes one by one, but the parent code is the role controller. I'm not sure what the conflict is here, if you are just objecting to where the go code lives in what package, or where it's called from, etc.

Reply:
>  We don't remove role labels from nodes. If you delete the role, it deletes everything it owns, including the node.

Are you sure we delete the node? I see that the node is owned by dataplane rather than dataplanerole[1].

> if you are just objecting to where the go code lives in what package

As I mentioned earlier I don't care where the code lives and how they're re-used, my concern is with the current design (i.e dependencies, ownership and how and when we reconcile), but let's move on. I've updated the baremetal provisioning PR to just hook baremetal provisioning in role reconcile[2].

[1] https://github.com/openstack-k8s-operators/dataplane-operator/blob/main/controllers/openstackdataplane_controller.go#L272
[2] https://github.com/openstack-k8s-operators/dataplane-operator/pull/165

Reply:
> > We don't remove role labels from nodes. If you delete the role, it deletes everything it owns, including the node.
> 
> Are you sure we delete the node? I see that the node is owned by dataplane rather than dataplanerole[1].

If created by the DataPlane, the node will be deleted when the DataPlane is deleted. If you delete the node on your own, it'd get re-created when the DataPlane is reconciled.

If you create Node and Role CR's outside of any DataPlane, and then delete one that needs the other, then you're going to get an undefined behavior, for which we need to add checks and error handling.

This is the same as if you have a Deployment with Volumes that reference a ConfigMap or Secret. If you then go in and delete that ConfigMap or Secret, you get an error, the next time the Deployment is reconciled you get an error.
 
> > if you are just objecting to where the go code lives in what package
> 
> As I mentioned earlier I don't care where the code lives and how they're re-used, my concern is with the current design (i.e dependencies, ownership and how and when we reconcile), but let's move on. I've updated the baremetal provisioning PR to just hook baremetal provisioning in role reconcile[2].

Please feel open to influence the current design to address your concerns. I'm happy to discuss any proposals or test out any PR's/prototypes.

Reply:
Splitting provision/config/deploy responsibilities over *ExternalCompute CRDs across multiple Node/Role/Datplane controllers could be indeed hard to implement, test, and maintain. We're not as much experienced as Deployment and ReplicaSet controllers authours. Given that, and added the recent "10,000 namespaces" scale testing feedback on etcd and k8s apiserver limitations, it might be architecture wise to ditch dataplane CRD (and even node CRDs) and scale nodes as replica counts of roles and its controllers.

Reply:
if we did that i dont think we would have a way to adopt osp 17.1 deployment since we could not model pre provisioned host anymore so i don't thinkthat will work unless we added a way to list pre-provision host in the role directly

basically a list of ansible connection triples (user, host, port)